File: raspberry-pi-ios-client/RaspberryPiClient/Data/Repository/Turtlesim/Spawn/TurtlesimServiceResponse.swift
//
//  TurtlesimServiceResponse.swift
//  RaspberryPiClient
//
//  Created by Shusuke Ota on 2025/12/28.
//

import Foundation

struct TurtlesimServiceResponse: RosServiceResponseValuesProtocol, RosServiceResponseCdrDecodable {
    let name: String

    static func decodeFromCdr(_ data: Data) -> TurtlesimServiceResponse? {
        guard let name = CdrHelpers.decodeCdrString(from: data, at: 4) else { return nil }
        return TurtlesimServiceResponse(name: name)
    }
}

File: raspberry-pi-android-client/app/src/main/java/com/shusuke/raspberry_pi_client/data/repository/turtlesim/TurtlesimServiceResponse.kt
package com.shusuke.raspberry_pi_client.data.repository.turtlesim

import com.shusuke.raspberry_pi_client.data.infrastructure.ros.cdr.CdrHelpers
import kotlinx.serialization.Serializable

/**
 * turtlesim の `/spawn` サービスのレスポンス
 */
@Serializable
data class TurtlesimServiceResponse(
    val name: String = ""
) {
    companion object {
        /**
         * CDR LE バイナリからデコードする。
         *
         * 構造: header(4) + name(CDR string)
         *
         * @return デコード成功時は [TurtlesimServiceResponse]、失敗時は null
         */
        fun decodeFromCdr(payload: ByteArray): TurtlesimServiceResponse? {
            if (payload.size < 4) return null
            val name = CdrHelpers.decodeCdrString(payload, 4) ?: ""
            return TurtlesimServiceResponse(name = name)
        }
    }
}
